src/hotspot/share/gc/serial/defNewGeneration.cpp

```diff
@@ -63,43 +63,75 @@
 #include "utilities/globalDefinitions.hpp"
 #include "utilities/stack.inline.hpp"
 
-//
-// DefNewGeneration functions.
+class IsAliveClosure: public BoolObjectClosure {
+  Generation* _young_gen;
+public:
+  IsAliveClosure(Generation* young_gen) : _young_gen(young_gen) {
+    assert(_young_gen->kind() == Generation::DefNew,
+        "Expected the young generation here");
+  }
 
-// Methods of protected closure types.
+  bool do_object_b(oop p) {
+    return cast_from_oop<HeapWord*>(p) >= _young_gen->reserved().end() || p->is_forwarded();
+  }
+};
 
-DefNewGeneration::IsAliveClosure::IsAliveClosure(Generation* young_gen) : _young_gen(young_gen) {
-  assert(_young_gen->kind() == Generation::DefNew, "Expected the young generation here");
-}
+class KeepAliveClosure: public OopClosure {
+  ScanWeakRefClosure* _cl;
+  CardTableRS* _rs;
+  HeapWord* _boundary;
 
-bool DefNewGeneration::IsAliveClosure::do_object_b(oop p) {
-  return cast_from_oop<HeapWord*>(p) >= _young_gen->reserved().end() || p->is_forwarded();
-}
+  template <class T>
+  inline void do_oop_work(T* p) {
+#ifdef ASSERT
+    {
+      // We never expect to see a null reference being processed
+      // as a weak reference.
+      oop obj = RawAccess<IS_NOT_NULL>::oop_load(p);
+      assert (oopDesc::is_oop(obj), "expected an oop while scanning weak refs");
+    }
+#endif // ASSERT
 
-DefNewGeneration::FastKeepAliveClosure::
-FastKeepAliveClosure(DefNewGeneration* g, ScanWeakRefClosure* cl) :
-  _cl(cl) {
-  _rs = GenCollectedHeap::heap()->rem_set();
-  _boundary = g->reserved().end();
-}
+    Devirtualizer::do_oop(_cl, p);
 
-void DefNewGeneration::FastKeepAliveClosure::do_oop(oop* p)       { DefNewGeneration::FastKeepAliveClosure::do_oop_work(p); }
-void DefNewGeneration::FastKeepAliveClosure::do_oop(narrowOop* p) { DefNewGeneration::FastKeepAliveClosure::do_oop_work(p); }
+    // Optimized for Defnew generation if it's the youngest generation:
+    // we set a younger_gen card if we have an older->youngest
+    // generation pointer.
+    oop obj = RawAccess<IS_NOT_NULL>::oop_load(p);
+    if ((cast_from_oop<HeapWord*>(obj) < _boundary) && GenCollectedHeap::heap()->is_in_reserved(p)) {
```

Comment:
I'm wondering whether the second part of the condition can ever be false. From what I can see, the `KeepAliveClosure` is only ever used during reference processing (afaict, may be wrong), and I can't think of a reason why the `Reference` would ever be outside the heap.
Feel free to ignore for this change, I can clean this up.